### PR TITLE
Add Redshift database support

### PR DESCRIPTION
**Pull request by pankajkoti**
# Description
## What is the current behavior?
No support available for Redshift database

## What is the new behavior?
Add support for Redshift database for the following operators:
- [x] load_file
- [x] transform
- [x] drop_table
- [x] run_raw_sql
- [x] append
- [x] export_file
- [x] dataframe

Note: The `merge` operator implementation needs some discussion and will be
provided in a subsequent PR.

## Does this introduce a breaking change?
No.

### Checklist
- [x] Created tests which fail without the change (if possible)

closes: #612

### CODE REVIEW

**Comment by pankajastro** [comment]
It would be great if can `squash` your commit.

File: .github/ci-test-connections.yaml [hunk 1]
```diff
@@ -52,3 +52,10 @@ connections:
     conn_type: aws
     description: null
     extra: null
+  - conn_id: redshift_conn
+    conn_type: redshift
+    schema: "dev"
+    host: "utkarsh-cluster.cdru7mxqmtyx.us-east-2.redshift.amazonaws.com"
```
Comment (pankajastro) [inline .github/ci-test-connections.yaml:58]: it is not so secret but shall we also take this also from env?

File: src/astro/databases/aws/redshift.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        return Metadata(database=database, schema=self.DEFAULT_SCHEMA)
+
+    def schema_exists(self, schema: str) -> bool:
+        """
+        Checks if a dataset exists in the Redshift
+
+        :param schema: Redshift namespace
+        """
+        schema_result = self.hook.run(
+            "SELECT schema_name FROM information_schema.schemata WHERE lower(schema_name) = lower(%s);",
```
Comment (pankajastro) [inline src/astro/databases/aws/redshift.py:66]: is `SELECT schema_name FROM information_schema.schemata WHERE lower(schema_name) = lower(%s);` return list having  1 row?

File: src/astro/databases/aws/redshift.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """Return SQAlchemy engine."""
+        uri = self.hook.get_uri()
+        return create_engine(uri)
+
+    @property
+    def default_metadata(self) -> Metadata:
+        """
+        Fill in default metadata values for table objects addressing redshift databases
+
+        :return:
```
Comment (pankajastro) [inline src/astro/databases/aws/redshift.py:53]: since at most of the places we don't have this in docstring so let remove it

File: src/astro/databases/aws/redshift.py (hunk 2 above)
Comment (pankajkoti) [reply src/astro/databases/aws/redshift.py:62]: Yes, it returns empty list when schema does not exist and otherwise ideally list of length 1 when schema exists. The logic is consistent to what we have for postgres database.

**Comment by pankajkoti** [comment]
> It would be great if can `squash` your commit.

Squashing looks difficult as the branch has been rebased multiple times and also few conflicts have been handled with merge commits from main time to time. However, when the branch is merged to main it will appear as a single commit.

File: src/astro/databases/aws/redshift.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
+"""AWS Redshift table implementation."""
+from typing import Any, Dict, List, Optional, Tuple
+
+import pandas as pd
+from airflow.providers.amazon.aws.hooks.redshift_sql import RedshiftSQLHook
+
+from google.api_core.exceptions import (
```
Comment (sunank200) [inline src/astro/databases/aws/redshift.py]: Is exceptions from google core api's relevant here?

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -124,6 +124,10 @@ jobs:
       AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
       AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
       AIRFLOW__ASTRO_SDK__SQL_SCHEMA: astroflow_ci
+      REDSHIFT_DATABASE: dev
+      REDSHIFT_HOST: utkarsh-cluster.cdru7mxqmtyx.us-east-2.redshift.amazonaws.com
```
Comment (sunank200) [inline .github/workflows/ci.yaml:128]: Should this URL be part of repo? It can be taken from secrets?

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -210,6 +214,10 @@ jobs:
...
```
Comment (sunank200) [inline .github/workflows/ci.yaml:218]: Should this URL be part of repo? It can be taken from secrets?

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -291,6 +299,10 @@ jobs:
...
```
Comment (sunank200) [inline .github/workflows/ci.yaml:303]: Should this URL be part of repo? It can be taken from secrets?

File: example_dags/example_load_file.py [hunk 8]
```diff
@@ -151,4 +151,28 @@
     )
     # [END load_file_example_12]
 
+    # [START load_file_example_13]
```
Comment (sunank200) [inline example_dags/example_load_file.py:154]: May be we should add one example of loading file from local as well as this DAG would test that scenario

File: src/astro/databases/aws/redshift.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    @property
+    def sqlalchemy_engine(self) -> Engine:
+        """Return SQAlchemy engine."""
+        uri = self.hook.get_uri()
+        return create_engine(uri)
+
+    @property
+    def default_metadata(self) -> Metadata:
+        """Fill in default metadata values for table objects addressing redshift databases"""
+        # TODO: Change airflow RedshiftSQLHook to fetch database and schema separately.
```
Comment (sunank200) [inline src/astro/databases/aws/redshift.py:51]: We should comment the limitation here from airflow RedshiftSQLHook side that it treats both schema and database in the same way.

Comment (sunank200) [inline src/astro/databases/aws/redshift.py:51]: We should comment the limitation here from airflow RedshiftSQLHook side that it treats both schema and database in the same way.

File: src/astro/databases/aws/redshift.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        database = self.hook.conn.schema
+        return Metadata(database=database, schema=self.DEFAULT_SCHEMA)
+
+    def schema_exists(self, schema: str) -> bool:
+        """
+        Checks if a dataset exists in the Redshift
+
+        :param schema: Redshift namespace
+        """
+        schema_result = self.hook.run(
```
Comment (sunank200) [inline src/astro/databases/aws/redshift.py:61]: What happens when schema_name is invalid or when this query fails?

File: tests/databases/test_redshift.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        (DEFAULT_CONN_ID, re.compile(r"^(redshift\+redshift_connector://)(.*)")),
+        (
+            CUSTOM_CONN_ID,
+            r"^(redshift\+redshift_connector://)(.*)(us-east-2.redshift.amazonaws.com:5439/dev)$",
+        ),
+    ],
+    ids=SUPPORTED_CONN_IDS,
+)
+def test_redshift_sqlalchemy_engine(conn_id, expected_uri):
+    """Test getting a redshift based sqla engine."""
```
Comment (sunank200) [inline tests/databases/test_redshift.py:48]:
```suggestion
    """Test getting a redshift based sqlalchemy engine."""
```

File: tests/databases/test_redshift.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+@pytest.mark.parametrize(
+    "database_table_fixture",
+    [
+        {"database": Database.REDSHIFT},
+    ],
+    indirect=True,
+    ids=["redshift"],
+)
+def test_load_file_to_table(database_table_fixture):
+    """Test loading on files to redshift database"""
```
Comment (sunank200) [inline tests/databases/test_redshift.py:156]: We should add integration test to load folder with a file and sub folder. That should run fine. Also this test should add cases to load from remote locations like S3 and GCS.

**Review by sunank200 — COMMENTED** [review]
Overall good work with implementation of adding Redshift @pankajkoti . I have added few comments.

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (pankajkoti) [reply .github/workflows/ci.yaml:128]: I was planning to keep it as part of secrets earlier. But, Kaxil mentioned that it is not a secret and can be taken as an environment variable.

File: src/astro/databases/aws/redshift.py (hunk 4 above)
Comment (pankajkoti) [reply src/astro/databases/aws/redshift.py]: Perhaps you're reviewed from some stale commit. This is not the case with the latest commit. https://github.com/astronomer/astro-sdk/pull/639/files#diff-b6bfa1674a3103f5a361d30fed6d115dca45133e0ad89b8d956382d63da12daa

File: .github/workflows/ci.yaml (hunk 6 above)
Comment (pankajkoti) [reply .github/workflows/ci.yaml:218]: I was planning to keep it as part of secrets earlier. But, Kaxil mentioned that it is not a secret and can be taken as an environment variable.

File: .github/workflows/ci.yaml (hunk 7 above)
Comment (pankajkoti) [reply .github/workflows/ci.yaml:303]: I was planning to keep it as part of secrets earlier. But, Kaxil mentioned that it is not a secret and can be taken as an environment variable.

File: src/astro/databases/aws/redshift.py (hunk 10 above)
Comment (pankajkoti) [reply src/astro/databases/aws/redshift.py:61]: If the schema does not exist, the base class `create_schema_if_needed` method takes care of creating the needed schema.

File: src/astro/databases/aws/redshift.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class RedshiftDatabase(BaseDatabase):
+    """
+    Handle interactions with Redshift databases.
+    """
+
+    DEFAULT_SCHEMA = REDSHIFT_SCHEMA
+
+    illegal_column_name_chars: List[str] = ["."]
```
Comment (utkarsharma2) [inline src/astro/databases/aws/redshift.py:26]: Do we know these chars are not valid col names in reshifts?

File: src/astro/databases/aws/redshift.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    illegal_column_name_chars_replacement: List[str] = ["_"]
```
Comment (utkarsharma2) [inline src/astro/databases/aws/redshift.py:27]: And this is acceptable in col name?

File: tests/databases/test_redshift.py [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+def test_redshift_run_sql():
+    """Test run_sql against redshift database"""
+    statement = "SELECT 1 + 1;"
+    database = RedshiftDatabase(conn_id=CUSTOM_CONN_ID)
+    response = database.run_sql(statement)
+    assert response.first()[0] == 2
+
+
+@pytest.mark.integration
+def test_table_exists_raises_exception():
```
Comment (utkarsharma2) [inline tests/databases/test_redshift.py:66]: @pankajastro Why is this test name with the suffix 'raises_exceptions'? I don't see an exception here.

File: tests/databases/test_redshift.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+    """Test loading on files to redshift database"""
+    database, target_table = database_table_fixture
+    filepath = str(pathlib.Path(CWD.parent, "data/sample.csv"))
+    database.load_file_to_table(File(filepath), target_table, {})
+
+    df = database.hook.get_pandas_df(
+        f"SELECT * FROM {database.get_table_qualified_name(target_table)}"
+    )
+    assert len(df) == 3
+    expected = pd.DataFrame(
```
Comment (utkarsharma2) [inline tests/databases/test_redshift.py:165]: @pankajkoti do we have guarantees of the ordering of rows? or should we explicitly sort rows?

Comment (utkarsharma2) [reply tests/databases/test_redshift.py:165]: in past, I have seen test cases breaking because the order of rows changed.

File: tests/databases/test_redshift.py [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+    ids=["redshift"],
+)
+def test_export_table_to_file_overrides_existing_file(database_table_fixture):
+    """
+    Test export_table_to_file_file() where the end file already exists,
+    should result in overriding the existing file
+    """
+    database, populated_table = database_table_fixture
+
+    filepath = str(pathlib.Path(CWD.parent, "data/sample.csv"))
```
Comment (utkarsharma2) [inline tests/databases/test_redshift.py:217]: @pankajkoti How do we know the file is replaced? I see the same file in the fixture and test getting loaded.

File: tests/databases/test_redshift.py (hunk 12 above)
Comment (pankajkoti) [reply tests/databases/test_redshift.py:156]: Addressed

File: tests/sql/operators/test_merge.py [hunk 18]
```diff
@@ -109,7 +109,10 @@ def run_merge(target_table: Table, source_table: Table, merge_parameters, mode):
 )
 @pytest.mark.parametrize(
     "database_table_fixture",
-    [{"database": Database.SNOWFLAKE}, {"database": Database.BIGQUERY}],
+    [
```
Comment (utkarsharma2) [inline tests/sql/operators/test_merge.py:112]: @pankajkoti  I think this is a formatting change, it should be part of a different PR, right?

File: tests/sql/operators/test_merge.py [hunk 19]
```diff
@@ -109,7 +109,10 @@ def run_merge(target_table: Table, source_table: Table, merge_parameters, mode):
...
+        {"database": Database.SNOWFLAKE},
+        {"database": Database.BIGQUERY},
+    ],
```
Comment (utkarsharma2) [inline tests/sql/operators/test_merge.py:115]: @pankajkoti We should have tests for Redshift as well right?

Comment (pankajkoti) [reply tests/sql/operators/test_merge.py:115]: It was decided that merge operator should be in a separate PR and hence the tests would go in that PR.

File: src/astro/databases/aws/redshift.py (hunk 13 above)
Comment (pankajkoti) [reply src/astro/databases/aws/redshift.py:26]: yes, that is correct, Redshift does not accept `.` column name characters

File: src/astro/databases/aws/redshift.py (hunk 14 above)
Comment (pankajkoti) [reply src/astro/databases/aws/redshift.py:27]: yes, this is acceptable.

File: tests/sql/operators/test_merge.py (hunk 18 above)
Comment (pankajkoti) [reply tests/sql/operators/test_merge.py:112]: I believe formatting change can go in this PR.

File: tests/databases/test_redshift.py (hunk 15 above)
Comment (pankajkoti) [reply tests/databases/test_redshift.py]: renamed the function, it was copied over from `test_postgres` which has the same check.

File: tests/databases/test_redshift.py (hunk 16 above)
Comment (pankajkoti) [reply tests/databases/test_redshift.py:165]: Sorted rows for all checks.

File: tests/databases/test_redshift.py (hunk 17 above)
Comment (pankajkoti) [reply tests/databases/test_redshift.py]: Right. Creating a new temporary file with different initial values and then overriding the file.

**Review by utkarsharma2 — APPROVED** [review]
LGTM

**Review by sunank200 — APPROVED** [review]
LGTM. Just deepsource and CI has to run fine
